// @locker/html-sanitizer/dist/index.cjs.js
/*!
 * Copyright (C) 2019 salesforce.com, inc.
 */
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var shared$LWS = require('@locker/shared');
var sharedDom$LWS = require('@locker/shared-dom');
var sharedUrl$LWS = require('@locker/shared-url');
var createDOMPurify$LWS = require('dompurify');
var trustedTypes$LWS = require('@locker/trusted-types');
function _interopDefaultLegacy$LWS(e$LWS) {
  return e$LWS && typeof e$LWS === 'object' && 'default' in e$LWS ? e$LWS : {
    'default': e$LWS
  };
}
var createDOMPurify__default$LWS = /*#__PURE__*/_interopDefaultLegacy$LWS(createDOMPurify$LWS);
const additionalAttributes$LWS = ['role', 'part', 'target'];
const htmlTags$LWS = ['a', 'abbr', 'acronym', 'address', 'area', 'article', 'aside', 'audio', 'b', 'bdi', 'bdo', 'big', 'blockquote', 'body', 'br', 'button', 'caption', 'canvas', 'center', 'cite', 'code', 'col', 'colgroup', 'command', 'datalist', 'dd', 'del', 'details', 'dfn', 'dir', 'div', 'dl', 'dt', 'em', 'fieldset', 'figure', 'figcaption', 'footer', 'form', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'head', 'header', 'hgroup', 'hr', 'i', 'iframe', 'img', 'input', 'ins', 'keygen', 'kbd', 'label', 'legend', 'li', 'map', 'mark', 'menu', 'meter', 'nav', 'ol', 'optgroup', 'option', 'output', 'p', 'pre', 'progress', 'q', 'rp', 'rt', 'ruby', 's', 'samp', 'section', 'select', 'small', 'source', 'span', 'strike', 'strong', 'style', 'sub', 'summary', 'sup', 'table', 'tbody', 'td', 'textarea', 'tfoot', 'th', 'thead', 'time', 'tr', 'track', 'tt', 'u', 'ul', 'var', 'video', 'wbr'];
const miscTags$LWS = ['#comment', '#document-fragment'];
const svgTags$LWS = ['svg', 'a', 'altglyph', 'altglyphdef', 'altglyphitem', 'animatecolor', 'animatemotion', 'animatetransform', 'audio', 'canvas', 'circle', 'clippath', 'defs', 'desc', 'ellipse', 'filter', 'font', 'g', 'glyph', 'glyphref', 'hkern', 'image', 'line', 'lineargradient', 'marker', 'mask', 'mpath', 'path', 'pattern', 'polygon', 'polyline', 'radialgradient', 'rect', 'stop', 'switch', 'symbol', 'text', 'textpath', 'title', 'tref', 'tspan', 'video', 'view', 'vkern', 'use'];
const allHTMLTags$LWS = shared$LWS.ArrayConcat(htmlTags$LWS, svgTags$LWS, miscTags$LWS);
const allSVGTags$LWS = shared$LWS.ArrayConcat(svgTags$LWS, miscTags$LWS);
const CUSTOM_ELEMENT_HANDLING$LWS = {
  attributeNameCheck: /.+/,
  allowCustomizedBuiltInElements: false,
  // The regexp is based on the WHATWG spec:
  // https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name
  // However, DOMPurify sanitizes unicode characters (\u0000-\uFFFF) in tag name.
  tagNameCheck: /^[a-z][-_.\w]*-[-.0-9_a-z\xB7\xC0-\xD6\xD8-\xF6\xF8-\u37D0\u37F-\u1FFF\u200C-\u200D\u203F-\u2040\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\u10000-\uEFFFF]*$/
};
// A generic config in which the sanitizer attempts in place sanitization and
// returns node.
const NODE_ALL_IN_PLACE$LWS = {
  ADD_ATTR: shared$LWS.shallowCloneArray(additionalAttributes$LWS),
  // Add '#document-fragment' to ALLOWED_TAGS to avoid a forbidden root node
  // exception.
  // https://github.com/cure53/DOMPurify/issues/664
  ALLOWED_TAGS: shared$LWS.shallowCloneArray(allHTMLTags$LWS),
  CUSTOM_ELEMENT_HANDLING: shared$LWS.ObjectAssign({}, CUSTOM_ELEMENT_HANDLING$LWS),
  IN_PLACE: true,
  // @ts-ignore this type is on DOMPurifyConfig
  TRUSTED_TYPES_POLICY: trustedTypes$LWS.trusted
};
// A config to use only svg tags in which the sanitizer returns a document
// fragment.
const NODE_SVG$LWS = {
  ADD_ATTR: shared$LWS.shallowCloneArray(additionalAttributes$LWS),
  ALLOWED_TAGS: shared$LWS.shallowCloneArray(allSVGTags$LWS),
  CUSTOM_ELEMENT_HANDLING: shared$LWS.ObjectAssign({}, CUSTOM_ELEMENT_HANDLING$LWS),
  RETURN_DOM_FRAGMENT: true,
  SANITIZE_DOM: false,
  // @ts-ignore this type is on DOMPurifyConfig
  TRUSTED_TYPES_POLICY: trustedTypes$LWS.trusted
};
// A config to use only tags allowed for blob and file.
const STRING_BLOB_HTML$LWS = {
  ADD_ATTR: shared$LWS.shallowCloneArray(additionalAttributes$LWS),
  ALLOWED_TAGS: shared$LWS.ReflectApply(shared$LWS.ArrayProtoFilter, allHTMLTags$LWS, [tag$LWS => tag$LWS !== 'iframe']),
  CUSTOM_ELEMENT_HANDLING: shared$LWS.ObjectAssign({}, CUSTOM_ELEMENT_HANDLING$LWS),
  SANITIZE_DOM: false,
  // @ts-ignore this type is on DOMPurifyConfig
  TRUSTED_TYPES_POLICY: trustedTypes$LWS.trusted
};
var CONFIG$LWS = /*#__PURE__*/Object.freeze({
  __proto__: null,
  NODE_ALL_IN_PLACE: NODE_ALL_IN_PLACE$LWS,
  NODE_SVG: NODE_SVG$LWS,
  STRING_BLOB_HTML: STRING_BLOB_HTML$LWS
});
const instancesBySandboxKeyRegistry$LWS = {
  __proto__: null
};
const SANITIZE_ATTRIBUTES_LIST$LWS = shared$LWS.toSafeArray(['href', 'xlink:href']);
const SHARED_SVG_SANITIZER_KEY$LWS = 'SHARED_SVG_SANITIZER_KEY';
shared$LWS.ReflectApply(sharedDom$LWS.DocumentProtoCreateElement, document, ['template']);
// Queue for managing pending XHR requests.
const queue$LWS = shared$LWS.toSafeSet(new shared$LWS.SetCtor());
// A regexp to find all non lowercase alphanumeric.
const urlReplacerRegExp$LWS = /[^a-z0-9]+/gi;
function createSantizerHooksRegistry$LWS(sandboxKey$LWS) {
  return {
    __proto__: null,
    // uponSanitizeAttribute is generic, so its definition can be a reused function
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    uponSanitizeAttribute: uponSanitizeAttribute$LWS,
    // uponSanitizeElement is sandbox-key-specific
    uponSanitizeElement(node$LWS, data$LWS, config$LWS) {
      var _config$CUSTOM_ELEMEN$LWS;
      const {
        tagName: tagName$LWS
      } = data$LWS;
      // istanbul ignore next: optional chaining and nullish coalescing results in an expansion that contains an unreachable "void 0" branch for every occurrence of the operator
      const tagNameCheck$LWS = config$LWS == null || (_config$CUSTOM_ELEMEN$LWS = config$LWS.CUSTOM_ELEMENT_HANDLING) == null ? void 0 : _config$CUSTOM_ELEMEN$LWS.tagNameCheck;
      if (tagNameCheck$LWS && shared$LWS.ReflectApply(shared$LWS.RegExpProtoTest, tagNameCheck$LWS, [tagName$LWS]) &&
      // This MUST be called ONLY after we've ensured the previous
      // two conditions have been met, otherwise it may result in an
      // unwanted "Illegal invocation" on a node we don't care about.
      !shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoHasAttribute, node$LWS, [sharedDom$LWS.CUSTOM_ELEMENT_REGISTRY_ATTRIBUTE_NAME])) {
        shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoSetAttribute, node$LWS, [sharedDom$LWS.CUSTOM_ELEMENT_REGISTRY_ATTRIBUTE_NAME, sandboxKey$LWS]);
      }
    }
  };
}
function createUrlContainer$LWS(url$LWS) {
  const container$LWS = shared$LWS.ReflectApply(sharedDom$LWS.DocumentProtoCreateElement, document, ['div']);
  shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoSetAttribute, container$LWS, ['style', 'display:none']);
  shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoSetAttribute, container$LWS, ['id', url$LWS]);
  const body$LWS = shared$LWS.ReflectApply(sharedDom$LWS.DocumentProtoBodyGetter, document, []);
  shared$LWS.ReflectApply(sharedDom$LWS.NodeProtoAppendChild, body$LWS, [container$LWS]);
  return container$LWS;
}
function getSanitizerForConfig$LWS(sandboxKey$LWS, configName$LWS) {
  if (typeof sandboxKey$LWS !== 'string') {
    throw new shared$LWS.LockerSecurityError(shared$LWS.ERR_INVALID_SANDBOX_KEY);
  }
  if (typeof configName$LWS !== 'string') {
    throw new shared$LWS.LockerSecurityError('Invalid config name.');
  }
  let sandboxConfiguredSanitizerInstances$LWS = instancesBySandboxKeyRegistry$LWS[sandboxKey$LWS];
  if (sandboxConfiguredSanitizerInstances$LWS === undefined) {
    sandboxConfiguredSanitizerInstances$LWS = instancesBySandboxKeyRegistry$LWS[sandboxKey$LWS] = {
      __proto__: null
    };
  }
  let configuredDOMPurifyInstance$LWS = sandboxConfiguredSanitizerInstances$LWS[configName$LWS];
  if (configuredDOMPurifyInstance$LWS === undefined) {
    const config$LWS = CONFIG$LWS[configName$LWS];
    configuredDOMPurifyInstance$LWS = createDOMPurify__default$LWS["default"]();
    configuredDOMPurifyInstance$LWS.setConfig(config$LWS);
    const hooksRegistry$LWS = createSantizerHooksRegistry$LWS(sandboxKey$LWS);
    for (const hookName$LWS in hooksRegistry$LWS) {
      configuredDOMPurifyInstance$LWS.addHook(hookName$LWS, hooksRegistry$LWS[hookName$LWS]);
    }
    sandboxConfiguredSanitizerInstances$LWS[configName$LWS] = configuredDOMPurifyInstance$LWS;
  }
  return configuredDOMPurifyInstance$LWS;
}
function sanitizeSvgTextReturnDOM$LWS(dirty$LWS) {
  const svgSanitizer$LWS = getSanitizerForConfig$LWS(SHARED_SVG_SANITIZER_KEY$LWS, 'NODE_SVG');
  return svgSanitizer$LWS.sanitize(dirty$LWS);
}
function sanitizeSvgHref$LWS(url$LWS) {
  const urlAsString$LWS = shared$LWS.toSafeStringValue(url$LWS);
  if (shared$LWS.ReflectApply(shared$LWS.StringProtoStartsWith, urlAsString$LWS, ['#'])) {
    return url$LWS;
  }
  const normalizedHref$LWS = parseHref$LWS(urlAsString$LWS);
  // Sanitize only for supported URL_SCHEMES_LIST.
  if (sharedUrl$LWS.URL_SCHEMES_LIST.includes(normalizedHref$LWS.protocol)) {
    const container$LWS = shared$LWS.ReflectApply(sharedDom$LWS.DocumentProtoGetElementById, document, [normalizedHref$LWS.normalizedURL]);
    // Have we sanitized this URL already?
    if (container$LWS && normalizedHref$LWS.normalizedFragment) {
      checkExistingAndDequeue$LWS(container$LWS, normalizedHref$LWS);
    } else if (!container$LWS) {
      fetchAndSanitize$LWS(normalizedHref$LWS);
    }
    // If this has been in the form of http://my-url/file.svg#fragment we
    // return the normalized fragment otherwise we return the normalized URL.
    return normalizedHref$LWS.requestedFragment ? `#${normalizedHref$LWS.normalizedFragment}` : `#${normalizedHref$LWS.normalizedURL}`;
  }
  return url$LWS;
}
function updater$LWS(container$LWS, normalizedHref$LWS) {
  const {
    normalizedFragment: normalizedFragment$LWS,
    requestedFragment: requestedFragment$LWS
  } = normalizedHref$LWS;
  let el$LWS = shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoQuerySelector, container$LWS, [`#${normalizedFragment$LWS}`]);
  // istanbul ignore else
  if (el$LWS === null) {
    try {
      el$LWS = shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoQuerySelector, container$LWS, [`#${requestedFragment$LWS}`]);
      shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoSetAttribute, el$LWS, ['id', normalizedFragment$LWS]);
    } catch (_unused$LWS) {
      // Catch all malformed CSS3 selectors.
      // getElementById not available on Node.
      // Cannot use document.getElementById because multiple containers
      // may have the same ids for SVG elements.
    }
  }
}
function checkExistingAndDequeue$LWS(container$LWS, normalizedHref$LWS) {
  if (queue$LWS.has(normalizedHref$LWS.normalizedURL)) {
    // Wait for request to finish, then update content.
    const interval$LWS = sharedDom$LWS.WindowSetInterval(() => {
      // istanbul ignore else
      if (!queue$LWS.has(normalizedHref$LWS.normalizedURL)) {
        updater$LWS(container$LWS, normalizedHref$LWS);
        sharedDom$LWS.WindowClearInterval(interval$LWS);
      }
    }, 50);
  } else {
    updater$LWS(container$LWS, normalizedHref$LWS);
  }
}
function fetchAndSanitize$LWS(normalizedHref$LWS) {
  // This is the first time we see this href.
  const container$LWS = createUrlContainer$LWS(normalizedHref$LWS.normalizedURL);
  // Put the URL we're fetching in a queue.
  queue$LWS.add(normalizedHref$LWS.normalizedURL);
  // Initiate an XHR to fetch the resource.
  const xhr$LWS = new sharedDom$LWS.XhrCtor();
  shared$LWS.ReflectApply(sharedDom$LWS.EventTargetProtoAddEventListener, xhr$LWS, ['load', () => {
    const status$LWS = shared$LWS.ReflectApply(sharedDom$LWS.XhrProtoStatusGetter, xhr$LWS, []);
    // istanbul ignore next: the following is well tested via integration-karma
    if (status$LWS === 200) {
      // Retrieved content should be sanitized immediately.
      const responseText$LWS = shared$LWS.ReflectApply(sharedDom$LWS.XhrProtoResponseTextGetter, xhr$LWS, []);
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      const fragment$LWS = sanitizeSvgTextReturnDOM$LWS(responseText$LWS);
      // Look for the container again in case other requests have finished
      // earlier for the same URL.
      if (normalizedHref$LWS.requestedFragment) {
        const el$LWS = shared$LWS.ReflectApply(sharedDom$LWS.DocumentFragmentProtoGetElementById, fragment$LWS, [normalizedHref$LWS.requestedFragment]);
        if (el$LWS) {
          shared$LWS.ReflectApply(sharedDom$LWS.ElementProtoSetAttribute, el$LWS, ['id', normalizedHref$LWS.normalizedFragment]);
        }
      }
      shared$LWS.ReflectApply(sharedDom$LWS.NodeProtoAppendChild, container$LWS, [fragment$LWS]);
      queue$LWS.delete(normalizedHref$LWS.normalizedURL);
    }
  }]);
  shared$LWS.ReflectApply(sharedDom$LWS.XhrProtoOpen, xhr$LWS, ['GET', normalizedHref$LWS.requestedURL]);
  shared$LWS.ReflectApply(sharedDom$LWS.XhrProtoSend, xhr$LWS, []);
}
function parseHref$LWS(url$LWS) {
  shared$LWS.ReflectApply(sharedDom$LWS.HTMLAnchorElementProtoHrefSetter, sharedUrl$LWS.normalizerAnchor, [url$LWS]);
  const href$LWS = shared$LWS.ReflectApply(sharedDom$LWS.HTMLAnchorElementProtoHrefGetter, sharedUrl$LWS.normalizerAnchor, []);
  const protocol$LWS = shared$LWS.ReflectApply(sharedDom$LWS.HTMLAnchorElementProtoProtocolGetter, sharedUrl$LWS.normalizerAnchor, []);
  const {
    0: requestedURL$LWS,
    1: requestedFragment$LWS
  } = shared$LWS.ReflectApply(shared$LWS.StringProtoSplit, href$LWS, ['#']);
  const loweredUrl$LWS = shared$LWS.ReflectApply(shared$LWS.StringProtoToLowerCase, requestedURL$LWS, []);
  // RegExp.prototype[Symbol.replace] resets the lastIndex of global regexp to 0.
  // https://tc39.es/ecma262/#sec-regexp.prototype-@@replace
  const normalizedURL$LWS = shared$LWS.ReflectApply(shared$LWS.StringProtoReplace, loweredUrl$LWS, [urlReplacerRegExp$LWS, '']);
  const normalizedFragment$LWS = requestedFragment$LWS ? `${normalizedURL$LWS}_${shared$LWS.ReflectApply(shared$LWS.StringProtoReplace, requestedFragment$LWS, [urlReplacerRegExp$LWS, ''])}` : '';
  return {
    normalizedFragment: normalizedFragment$LWS,
    normalizedURL: normalizedURL$LWS,
    protocol: protocol$LWS,
    requestedFragment: requestedFragment$LWS,
    requestedURL: requestedURL$LWS
  };
}
// Sanitize a URL representing a SVG href attribute value.
function uponSanitizeAttribute$LWS(node$LWS, data$LWS, _config$LWS) {
  const {
    attrValue: attrValue$LWS,
    attrName: attrName$LWS
  } = data$LWS;
  if (attrValue$LWS && shared$LWS.ReflectApply(shared$LWS.StringProtoToUpperCase, shared$LWS.ReflectApply(sharedDom$LWS.NodeProtoNodeNameGetter, node$LWS, []), []) === 'USE' && SANITIZE_ATTRIBUTES_LIST$LWS.includes(attrName$LWS)) {
    data$LWS.attrValue = sanitizeSvgHref$LWS(attrValue$LWS);
  }
  // To support Lit, we must tell DOMPurify that attributes starting with "@", ".", or "?" are allowed.
  // Ref:
  // https://lit.dev/docs/components/events/
  // https://lit.dev/docs/templates/expressions/#property-expressions
  // https://lit.dev/docs/templates/expressions/#boolean-attribute-expressions
  // istanbul ignore next: this is tested under all normal CI runs, but is not included in coverage
  if (attrName$LWS && (shared$LWS.ReflectApply(shared$LWS.StringProtoStartsWith, attrName$LWS, ['@']) || shared$LWS.ReflectApply(shared$LWS.StringProtoStartsWith, attrName$LWS, ['.']) || shared$LWS.ReflectApply(shared$LWS.StringProtoStartsWith, attrName$LWS, ['?']))) {
    data$LWS.forceKeepAttr = true;
  }
  return data$LWS;
}
function blobSanitizer$LWS(sandboxKey$LWS) {
  if (typeof sandboxKey$LWS !== 'string') {
    throw new shared$LWS.LockerSecurityError(shared$LWS.ERR_INVALID_SANDBOX_KEY);
  }
  return getSanitizerForConfig$LWS(sandboxKey$LWS, 'STRING_BLOB_HTML');
}
exports.blobSanitizer = blobSanitizer$LWS;
exports.getSanitizerForConfig = getSanitizerForConfig$LWS;
exports.sanitizeSvgHref = sanitizeSvgHref$LWS;
exports.sanitizeSvgTextReturnDOM = sanitizeSvgTextReturnDOM$LWS;
exports.uponSanitizeAttribute = uponSanitizeAttribute$LWS;
/*! version: 0.23.6 */
