// @lwrjs/core/build/cjs/middleware.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/core/src/middleware.ts
__markAsModule(exports);
__export(exports, {
  assetMiddleware: () => import_asset_middleware.assetMiddleware,
  bundleMiddleware: () => import_bundle_middleware.bundleMiddleware,
  hmrMiddleware: () => import_hmr_middleware.hmrMiddleware,
  localeMiddleware: () => import_locale_middleware.localeMiddleware,
  mappingMiddleware: () => import_mapping_middleware.mappingMiddleware,
  moduleMiddleware: () => import_module_middleware.moduleMiddleware,
  requestProcessorMiddleware: () => import_request_processor_middleware.requestProcessorMiddleware,
  resourceMiddleware: () => import_resource_middleware.resourceMiddleware,
  viewMiddleware: () => import_view_middleware.viewMiddleware
});
var import_request_processor_middleware = __toModule(require("./middleware/request-processor-middleware.cjs"));
var import_locale_middleware = __toModule(require("./middleware/locale-middleware.cjs"));
var import_hmr_middleware = __toModule(require("./middleware/hmr-middleware.cjs"));
var import_module_middleware = __toModule(require("./middleware/module-middleware.cjs"));
var import_bundle_middleware = __toModule(require("./middleware/bundle-middleware.cjs"));
var import_mapping_middleware = __toModule(require("./middleware/mapping-middleware.cjs"));
var import_asset_middleware = __toModule(require("./middleware/asset-middleware.cjs"));
var import_view_middleware = __toModule(require("./middleware/view-middleware.cjs"));
var import_resource_middleware = __toModule(require("./middleware/resource-middleware.cjs"));
