// @lwrjs/core/build/cjs/middleware/mapping-middleware.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/core/src/middleware/mapping-middleware.ts
__markAsModule(exports);
__export(exports, {
  mappingMiddleware: () => mappingMiddleware
});
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_instrumentation = __toModule(require("@lwrjs/instrumentation"));
var import_identity = __toModule(require("./utils/identity.cjs"));
var import_error_handling = __toModule(require("./utils/error-handling.cjs"));
function createMappingMiddleware(context) {
  const {
    appConfig,
    moduleBundler,
    moduleRegistry,
    runtimeEnvironment: defaultRuntimeEnvironment
  } = context;
  return async (req, res) => {
    if (!req.validateEnvironmentRequest(appConfig)) {
      res.status(400);
      res.send(import_diagnostics.descriptions.UNRESOLVABLE.INVALID_ENVIRONMENT(req.params.environment).message);
      return;
    }
    const {runtimeEnvironment, runtimeParams} = req.getRuntimeContext(defaultRuntimeEnvironment);
    const {moduleIds} = (0, import_identity.getMappingIdentity)(req);
    const importMetadata = await (0, import_instrumentation.getTracer)().trace({
      name: import_instrumentation.RequestHandlerSpan.GetMapping,
      attributes: {
        url: req.originalUrl
      }
    }, () => {
      return (0, import_shared_utils.getImportMetadataMappings)(moduleIds, runtimeEnvironment, runtimeParams, moduleRegistry, moduleBundler);
    });
    if (req.isSiteGeneration()) {
      res.setSiteGenerationMetadata({importMetadata});
    }
    res.status(200).type("application/json").send(importMetadata);
  };
}
function mappingMiddleware(app, context) {
  app.get([
    `/:apiVersion/mapping/:format/:compat/l/:locale/e/:environment/bi/:bundleSpecifier/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/l/:locale/e/:environment/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/l/:locale/bi/:bundleSpecifier/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/e/:environment/bi/:bundleSpecifier/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/l/:locale/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/e/:environment/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/bi/:bundleSpecifier/mp/:specifiers`,
    `/:apiVersion/mapping/:format/:compat/mp/:specifiers`
  ], (0, import_error_handling.handleErrors)(createMappingMiddleware(context)));
}
