// @lwrjs/core/build/es/middleware/bundle-middleware.js
import { LATEST_SIGNATURE } from '@lwrjs/shared-utils';
import { descriptions } from '@lwrjs/diagnostics';
import { RequestHandlerSpan, getTracer } from '@lwrjs/instrumentation';
import { getRequestImporter } from './utils/request.js';
import { getModuleIdentity } from './utils/identity.js';
import { normalizeResolvedUris } from './utils/metadata.js';
import { createUnsignedBundleRedirect } from './redirects/unsigned-module-redirect.js';
import { handleErrors } from './utils/error-handling.js';
function createBundleMiddleware(context) {
    const { appConfig, moduleRegistry, moduleBundler, runtimeEnvironment: defaultRuntimeEnvironment, } = context;
    const unsignedBundleRedirect = createUnsignedBundleRedirect(moduleBundler);
    return async (req, res) => {
        if (!req.validateEnvironmentRequest(appConfig)) {
            res.status(400);
            res.send(descriptions.UNRESOLVABLE.INVALID_ENVIRONMENT(req.params.environment).message);
            return;
        }
        if (!req.validateJsonRequest()) {
            res.status(400);
            res.send(descriptions.UNRESOLVABLE.INVALID_JSON().message);
            return;
        }
        if (!req.validateApiVersion(appConfig)) {
            res.status(400);
            res.send(descriptions.UNRESOLVABLE.INVALID_API_VERSION(req.params.apiVersion, appConfig.apiVersion)
                .message);
            return;
        }
        const { runtimeEnvironment, runtimeParams } = req.getRuntimeContext(defaultRuntimeEnvironment);
        const importer = req.query.importer
            ? await getRequestImporter(req, moduleRegistry, runtimeParams)
            : undefined;
        const { moduleId, signature } = getModuleIdentity(req, importer);
        if (moduleId.importer || !signature) {
            await unsignedBundleRedirect(req, res, moduleId, runtimeEnvironment, runtimeParams);
            return;
        }
        const sourceMapUrl = req.path.replace('/bundle/', '/sourcemaps/bundle/');
        const bundleDefinition = await getTracer().trace({
            name: RequestHandlerSpan.GetBundle,
            attributes: {
                specifier: moduleId.specifier,
                url: req.originalUrl,
            },
        }, () => {
            return moduleBundler.getModuleBundle(moduleId, 
            // bundle must be `true` to properly resolve bundles in amd
            { ...runtimeEnvironment, bundle: true, sourceMapUrl }, runtimeParams);
        });
        if (req.isSiteGeneration()) {
            res.setSiteGenerationMetadata({
                moduleDefinition: bundleDefinition,
                resolvedUris: await normalizeResolvedUris(bundleDefinition, runtimeEnvironment, runtimeParams, moduleBundler, moduleRegistry),
            });
        }
        if (signature !== LATEST_SIGNATURE) {
            res.setHeader('Cache-control', 'public, max-age=31536000, immutable');
        }
        res.status(200)
            .type('application/javascript')
            .send(await bundleDefinition.getCode());
    };
}
function createSourceMapMiddleware(context) {
    const { appConfig, moduleBundler, runtimeEnvironment: defaultRuntimeEnvironment } = context;
    return async (req, res) => {
        if (!req.validateEnvironmentRequest(appConfig)) {
            res.status(400);
            res.send(descriptions.UNRESOLVABLE.INVALID_ENVIRONMENT(req.params.environment).message);
            return;
        }
        const { runtimeEnvironment } = req.getRuntimeContext(defaultRuntimeEnvironment);
        const { moduleId, signature } = getModuleIdentity(req);
        const bundleDef = await moduleBundler.getModuleBundle(moduleId, runtimeEnvironment);
        if (signature !== LATEST_SIGNATURE) {
            res.setHeader('Cache-control', 'public, max-age=31536000, immutable');
        }
        res.status(200).type('application/json').send(bundleDef.map);
    };
}
export function bundleMiddleware(app, context) {
    app.get([
        `/:apiVersion/bundle/:format/:compat?/l/:locale/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/l/:locale/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/l/:locale/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/l/:locale/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/bundle/:format/:compat?/l/:locale/e/:environment/bi/:bundleSpecifier/module/mi/:specifier`,
        `/:apiVersion/bundle/:format/:compat?/l/:locale/bi/:bundleSpecifier/module/mi/:specifier`,
        `/:apiVersion/bundle/:format/:compat?/e/:environment/bi/:bundleSpecifier/module/mi/:specifier`,
        `/:apiVersion/bundle/:format/:compat?/bi/:bundleSpecifier/module/mi/:specifier`,
    ], handleErrors(createBundleMiddleware(context)));
    app.get([
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/l/:locale/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/l/:locale/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/l/:locale/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/l/:locale/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/e/:environment/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/bi/:bundleSpecifier/module/mi/:specifier/latest/:prettyUrl?`,
        `/:apiVersion/sourcemaps/bundle/:format/:compat?/bi/:bundleSpecifier/module/mi/:specifier/s/:signature/:prettyUrl?`,
    ], handleErrors(createSourceMapMiddleware(context)));
}
