// @lwrjs/server/build/cjs/koa/koa-request.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/server/src/koa/koa-request.ts
__markAsModule(exports);
__export(exports, {
  default: () => koa_request_default
});
var import_abstract_request = __toModule(require("../common/abstract-request.cjs"));
var KoaRequest = class extends import_abstract_request.AbstractMiddlewareRequest {
  constructor(ctx, isSiteGenerationRequest) {
    super(isSiteGenerationRequest);
    this.ctx = ctx;
  }
  get params() {
    return this.ctx.params || {};
  }
  get originalUrl() {
    return this.ctx.request.originalUrl;
  }
  get path() {
    return this.ctx.request.path;
  }
  get query() {
    return this.ctx.request.query;
  }
  get protocol() {
    return this.ctx.protocol;
  }
  get body() {
    return this.ctx.body;
  }
  get method() {
    return this.ctx.method;
  }
  get headers() {
    return this.ctx.headers;
  }
  get basePath() {
    return this.ctx.basePath || "";
  }
  set basePath(bp) {
    this.ctx.basePath = bp;
  }
  get locale() {
    return this.ctx.locale;
  }
  set locale(l) {
    this.ctx.locale = l;
  }
  acceptsLanguages() {
    return this.ctx.request.acceptsLanguages();
  }
  get(field) {
    return this.ctx.get(field);
  }
  cookie(key) {
    return this.ctx.cookies.get(key);
  }
};
var koa_request_default = KoaRequest;
