// @lwrjs/server/build/cjs/common/abstract-request.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/server/src/common/abstract-request.ts
__markAsModule(exports);
__export(exports, {
  AbstractMiddlewareRequest: () => AbstractMiddlewareRequest
});
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
function parseForwardedHeader(forwarded) {
  if (!forwarded)
    return void 0;
  const couples = forwarded.toLocaleLowerCase().split(";");
  const map = couples.reduce((m, c) => {
    const [key, value] = c.trim().split("=");
    m[key] = value;
    return m;
  }, {});
  const proto = map.proto ? `${map.proto}://` : "";
  return map.host ? `${proto}${map.host}` : void 0;
}
var AbstractMiddlewareRequest = class {
  constructor(isSiteGenerationRequest) {
    this.isSiteGenerationRequest = isSiteGenerationRequest || false;
  }
  cookie(key) {
    const rawCookies = (this.headers.cookie?.split(";") || []).map((c) => c.trim());
    const rawCookie = rawCookies.find((c) => c.startsWith(`${key}=`));
    return rawCookie ? decodeURIComponent(rawCookie.replace(`${key}=`, "")) : void 0;
  }
  isSiteGeneration() {
    return this.isSiteGenerationRequest;
  }
  hasJsonParam() {
    return this.query.json !== void 0;
  }
  isJsonRequest() {
    const acceptHeader = this.headers?.accept || "";
    const hasJsonHeader = acceptHeader.indexOf("application/json") >= 0;
    return hasJsonHeader || this.query.json !== void 0;
  }
  validateJsonRequest() {
    const hasJsonQueryParam = this.hasJsonParam();
    const hasAcceptHeader = this.headers?.accept !== void 0;
    const hasJsonAcceptHeader = hasAcceptHeader && this.headers?.accept?.includes("application/json");
    return !hasAcceptHeader && !hasJsonQueryParam || hasJsonQueryParam !== hasAcceptHeader || hasJsonQueryParam === hasJsonAcceptHeader;
  }
  validateEnvironmentRequest(config) {
    const targetEnvironment = this.params.environment;
    const defaultEnvironment = config.environment.default;
    const supportedEnvironments = config.environment.supported || [];
    return !targetEnvironment || targetEnvironment === defaultEnvironment || supportedEnvironments?.includes(targetEnvironment);
  }
  validateApiVersion(config) {
    const apiVersion = this.params.apiVersion;
    return apiVersion === config.apiVersion;
  }
  getRuntimeContext(defaultRuntimeEnvironment) {
    const {
      format,
      apiVersion,
      basePath: basePathParam,
      compat,
      locale: localeParam,
      environment,
      bundleSpecifier
    } = this.params;
    const debug = this.query.debug !== void 0;
    const bundle = !!bundleSpecifier;
    const runtimeEnvironment = {
      ...defaultRuntimeEnvironment,
      format: format || defaultRuntimeEnvironment.format,
      compat: compat || defaultRuntimeEnvironment.compat,
      debug: debug || defaultRuntimeEnvironment.debug,
      apiVersion: apiVersion || defaultRuntimeEnvironment.apiVersion,
      bundle: bundle || defaultRuntimeEnvironment.bundle,
      basePath: this.basePath || basePathParam || defaultRuntimeEnvironment.basePath
    };
    let host = parseForwardedHeader(this.headers?.forwarded) || this.headers?.host || void 0;
    if (host && !host.startsWith("http")) {
      host = `${this.headers?.["x-forwarded-proto"] || this.protocol}://${host}`;
    }
    const requestDepth = (0, import_shared_utils.parseRequestDepth)(this.headers, this.query) + 1;
    const basePath = runtimeEnvironment.basePath;
    const defaultLocale = runtimeEnvironment.i18n.defaultLocale;
    const locale = localeParam || this.locale || defaultLocale;
    const uiBasePath = runtimeEnvironment.i18n.uriPattern === "path-prefix" && defaultLocale !== locale ? `${basePath}/${locale}` : `${basePath}`;
    const assetBasePath = `${runtimeEnvironment.basePath}${runtimeEnvironment.defaultAssetPath}`;
    const runtimeParams = {
      basePath,
      locale,
      assetBasePath,
      uiBasePath,
      environment,
      host,
      requestDepth
    };
    return {
      runtimeEnvironment,
      runtimeParams
    };
  }
  getCoreProxy(appCoreProxy, proxyForSSR = false) {
    const origin = process.env.CORE_ORIGIN;
    const host = process.env.CORE_HOST;
    const servername = process.env.CORE_SERVERNAME;
    const override = process.env.PROXY_FOR_SSR;
    if (override === "false" || !proxyForSSR && override !== "true")
      return void 0;
    return origin ? {origin, host, servername} : appCoreProxy;
  }
};
