// @lwrjs/lwc-module-provider/build/cjs/cache.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/lwc-module-provider/src/cache.ts
__markAsModule(exports);
__export(exports, {
  DEFAULT_CACHE_FOLDER: () => DEFAULT_CACHE_FOLDER,
  DEFAULT_CACHE_INDEX: () => DEFAULT_CACHE_INDEX,
  DEFAULT_COMPILED_DIR: () => DEFAULT_COMPILED_DIR,
  addCompiledModuleCacheEntry: () => addCompiledModuleCacheEntry,
  getCompiledModuleCacheEntry: () => getCompiledModuleCacheEntry,
  setupModuleCache: () => setupModuleCache
});
var import_path = __toModule(require("path"));
var import_fs_extra = __toModule(require("fs-extra"));
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var DEFAULT_COMPILED_DIR = "lwc_compiled_modules";
var DEFAULT_CACHE_FOLDER = "cache";
var DEFAULT_CACHE_INDEX = `compiled.json`;
var NORMALIZE_PATH_REGEX = /[@\/#\.\?<>\\:\*\|"]/gm;
function setupModuleCache(cacheDir) {
  const lwcCacheDir = import_path.default.join(cacheDir, `${DEFAULT_COMPILED_DIR}`);
  const lwcCacheIndexPath = import_path.default.join(lwcCacheDir, DEFAULT_CACHE_INDEX);
  import_fs_extra.default.mkdirSync(`${lwcCacheDir}/${DEFAULT_CACHE_FOLDER}`, {recursive: true});
  if (!import_fs_extra.default.existsSync(lwcCacheIndexPath)) {
    import_fs_extra.default.writeFileSync(lwcCacheIndexPath, "[]");
    return {lwcCacheDir, lwcCacheIndex: new Map()};
  } else {
    const cacheIndexJson = import_fs_extra.default.readJSONSync(lwcCacheIndexPath, "utf-8");
    return {lwcCacheDir, lwcCacheIndex: new Map(cacheIndexJson)};
  }
}
var IndexPersister = class {
  async persist({lwcCacheIndex, lwcCacheDir}) {
    if (this.updateTimer) {
      clearTimeout(this.updateTimer);
    }
    this.lwcCacheIndex = lwcCacheIndex;
    this.lwcCacheDir = lwcCacheDir;
    this.updateTimer = setTimeout(() => {
      this.persistData();
      this.lwcCacheIndex = void 0;
      this.lwcCacheDir = void 0;
      this.updateTimer = void 0;
    }, 1e3);
  }
  persistData() {
    const lwcCacheIndexPath = import_path.default.join(this.lwcCacheDir, DEFAULT_CACHE_INDEX);
    import_fs_extra.default.writeJSONSync(lwcCacheIndexPath, [...this.lwcCacheIndex]);
    import_diagnostics.logger.debug("lwc-module-provider", `LWC Index Updated`);
  }
};
var INDEX_PERSISTER = new IndexPersister();
function getEnvVariant(targetSSR) {
  return targetSSR ? "server" : "client";
}
function getCacheKey(specifier, version, targetSSR) {
  return `${specifier}@${version}|${getEnvVariant(targetSSR)}`;
}
var moduleCount = 0;
async function addCompiledModuleCacheEntry(moduleSource, compilerResult, targetSSR, {lwcCacheIndex, lwcCacheDir}) {
  const {specifier, version, ownHash} = moduleSource;
  const cacheKey = getCacheKey(specifier, version, targetSSR);
  if (++moduleCount % 1e3 === 0) {
    import_diagnostics.logger.debug({label: `lwc-module-provider`, message: `Compiled Modules >${moduleCount}`});
  }
  const normalizedSpecifier = specifier.replace(NORMALIZE_PATH_REGEX, "_");
  const normalizedVersion = version.replace(NORMALIZE_PATH_REGEX, "_");
  const ssrVariant = getEnvVariant(targetSSR);
  const moduleFileName = `${normalizedSpecifier}_${normalizedVersion}_${ssrVariant}.js`;
  const cachedModulePath = import_path.default.join(lwcCacheDir, DEFAULT_CACHE_FOLDER, moduleFileName);
  const writeModulePromise = import_fs_extra.default.writeFile(cachedModulePath, compilerResult.code);
  const moduleMetaFileName = `${normalizedSpecifier}_${normalizedVersion}_${ssrVariant}.meta.json`;
  const cachedMetaPath = import_path.default.join(lwcCacheDir, DEFAULT_CACHE_FOLDER, moduleMetaFileName);
  const writeMetadataPromise = import_fs_extra.default.writeJSON(cachedMetaPath, compilerResult.metadata);
  lwcCacheIndex.set(cacheKey, {
    ownHash,
    module: `./${DEFAULT_CACHE_FOLDER}/${moduleFileName}`,
    moduleMeta: `./${DEFAULT_CACHE_FOLDER}/${moduleMetaFileName}`
  });
  const writeMetadata = INDEX_PERSISTER.persist({lwcCacheIndex, lwcCacheDir});
  return Promise.all([writeModulePromise, writeMetadataPromise, writeMetadata]);
}
function getCompiledModuleCacheEntry({specifier, version, ownHash}, targetSSR, {lwcCacheIndex, lwcCacheDir}) {
  const cacheKey = getCacheKey(specifier, version, targetSSR);
  const cacheEntry = lwcCacheIndex.get(cacheKey);
  if (!cacheEntry) {
    return;
  }
  const cacheModulePath = import_path.default.join(lwcCacheDir, cacheEntry.module);
  const cacheMetaPath = import_path.default.join(lwcCacheDir, cacheEntry.moduleMeta);
  if (ownHash === cacheEntry.ownHash && import_fs_extra.default.existsSync(cacheModulePath) && import_fs_extra.default.existsSync(cacheMetaPath)) {
    return {
      code: import_fs_extra.default.readFileSync(cacheModulePath, "utf-8"),
      metadata: JSON.parse(import_fs_extra.default.readFileSync(cacheMetaPath, "utf-8"))
    };
  }
}
