// @lwrjs/static/build/cjs/site-metadata.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/static/src/site-metadata.ts
__markAsModule(exports);
__export(exports, {
  SITE_LOCALE_PREFIX: () => SITE_LOCALE_PREFIX,
  SITE_SSR_PREFIX: () => SITE_SSR_PREFIX,
  SITE_VERSION_PREFIX: () => SITE_VERSION_PREFIX,
  SiteMetadataImpl: () => SiteMetadataImpl,
  getSiteBundleId: () => getSiteBundleId,
  getSiteResourceId: () => getSiteResourceId,
  parseSiteId: () => parseSiteId,
  resolveStaticBundleVersion: () => resolveStaticBundleVersion
});
var import_path = __toModule(require("path"));
var import_fs_extra = __toModule(require("fs-extra"));
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_decision_tree = __toModule(require("./utils/decision-tree.cjs"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var SITE_METADATA_PATH = ".metadata";
var STATIC_BUNDLE_METADATA_PATH = import_path.default.join(SITE_METADATA_PATH, "/bundle-metadata.json");
var DEBUG_STATIC_BUNDLE_METADATA_PATH = import_path.default.join(SITE_METADATA_PATH, "/bundle-metadata-debug.json");
var STATIC_RESOURCE_METADATA_PATH = import_path.default.join(SITE_METADATA_PATH, "/resource-metadata.json");
var DEBUG_STATIC_RESOURCE_METADATA_PATH = import_path.default.join(SITE_METADATA_PATH, "/resource-metadata-debug.json");
var STATIC_ASSET_METADATA_PATH = import_path.default.join(SITE_METADATA_PATH, "/asset-metadata.json");
var SITE_VERSION_PREFIX = `|${import_shared_utils.VERSION_SIGIL}/`;
var SITE_LOCALE_PREFIX = `|${import_shared_utils.LOCALE_SIGIL}/`;
var SITE_SSR_PREFIX = `|${import_shared_utils.SSR_SIGIL}`;
var SiteMetadataImpl = class {
  constructor(options) {
    this.options = options;
    this.siteBundles = this.readStaticBundleMetadata(options.rootDir, STATIC_BUNDLE_METADATA_PATH);
    this.debugSiteBundles = this.readStaticBundleMetadata(options.rootDir, DEBUG_STATIC_BUNDLE_METADATA_PATH);
    this.siteResources = this.readStaticResourceMetadata(options.rootDir, STATIC_RESOURCE_METADATA_PATH);
    this.debugSiteResources = this.readStaticResourceMetadata(options.rootDir, DEBUG_STATIC_RESOURCE_METADATA_PATH);
    this.siteAssets = this.readStaticAssetsMetadata(options.rootDir, STATIC_ASSET_METADATA_PATH);
  }
  getSiteRootDir() {
    return this.options.rootDir;
  }
  getSiteBundles() {
    return this.siteBundles;
  }
  getDebugSiteBundles() {
    return this.debugSiteBundles;
  }
  getSiteResources() {
    return this.siteResources;
  }
  getDebugSiteResources() {
    return this.debugSiteResources;
  }
  getSiteAssets() {
    return this.siteAssets;
  }
  getSiteBundlesDecisionTree() {
    if (!this.bundleDecisionTree) {
      this.bundleDecisionTree = new import_decision_tree.default();
      const localeFallbacks = (0, import_decision_tree.createFallbackMap)(this.options.i18n);
      for (const [key, bundle] of Object.entries(this.siteBundles.bundles)) {
        this.bundleDecisionTree.insert(key, bundle, false, localeFallbacks);
      }
      for (const [key, bundle] of Object.entries(this.debugSiteBundles.bundles)) {
        this.bundleDecisionTree.insert(key, bundle, true, localeFallbacks);
      }
    }
    return this.bundleDecisionTree;
  }
  getSiteResourcesDecisionTree() {
    if (!this.resourceDecisionTree) {
      this.resourceDecisionTree = new import_decision_tree.default();
      for (const [key, resource] of Object.entries(this.siteResources.resources)) {
        this.resourceDecisionTree.insert(key, resource, false);
      }
      for (const [key, resource] of Object.entries(this.debugSiteResources.resources)) {
        this.resourceDecisionTree.insert(key, resource, true);
      }
    }
    return this.resourceDecisionTree;
  }
  async persistSiteMetadata() {
    const siteMetadataPath = import_path.default.join(this.options.rootDir, SITE_METADATA_PATH);
    if (siteMetadataPath.indexOf("__skip_directory_creation__") !== -1)
      return;
    try {
      if (!import_fs_extra.default.existsSync(siteMetadataPath)) {
        await import_fs_extra.default.mkdir(siteMetadataPath, {recursive: true});
      }
      const bundleMetadataPath = import_path.default.join(this.options.rootDir, STATIC_BUNDLE_METADATA_PATH);
      await import_fs_extra.default.writeJSON(bundleMetadataPath, this.siteBundles, {spaces: 2});
      const debugBundleMetadataPath = import_path.default.join(this.options.rootDir, DEBUG_STATIC_BUNDLE_METADATA_PATH);
      await import_fs_extra.default.writeJSON(debugBundleMetadataPath, this.debugSiteBundles, {spaces: 2});
      const resourceMetadataPath = import_path.default.join(this.options.rootDir, STATIC_RESOURCE_METADATA_PATH);
      await import_fs_extra.default.writeJSON(resourceMetadataPath, this.siteResources, {spaces: 2});
      const debugResourceMetadataPath = import_path.default.join(this.options.rootDir, DEBUG_STATIC_RESOURCE_METADATA_PATH);
      await import_fs_extra.default.writeJSON(debugResourceMetadataPath, this.debugSiteResources, {spaces: 2});
      const assetMetadataPath = import_path.default.join(this.options.rootDir, STATIC_ASSET_METADATA_PATH);
      return import_fs_extra.default.writeJSON(assetMetadataPath, this.siteAssets, {spaces: 2});
    } catch (err) {
      import_diagnostics.logger.error(`[SiteMetadata] Failed to save site metadata ${siteMetadataPath}`);
      import_diagnostics.logger.error(err);
    }
  }
  readStaticBundleMetadata(staticRoot, metadataPath) {
    let bundleMetadataPath;
    let siteBundles = {bundles: {}};
    try {
      bundleMetadataPath = import_path.default.join(staticRoot, metadataPath);
      const savedMetadata = import_fs_extra.default.readJSONSync(bundleMetadataPath);
      siteBundles = savedMetadata;
    } catch (error) {
      if (error.code === "ENOENT") {
        import_diagnostics.logger.debug({
          label: `SiteMetadata`,
          message: `Failed to load Static Bundle Metadata: ${bundleMetadataPath}`
        });
      } else {
        throw error;
      }
    }
    return siteBundles;
  }
  readStaticResourceMetadata(staticRoot, metadataPath) {
    let resourceMetadataPath;
    let siteResources = {resources: {}};
    try {
      resourceMetadataPath = import_path.default.join(staticRoot, metadataPath);
      const savedMetadata = import_fs_extra.default.readJSONSync(resourceMetadataPath);
      siteResources = savedMetadata;
    } catch (error) {
      if (error.code === "ENOENT") {
        import_diagnostics.logger.debug({
          label: `SiteMetadata`,
          message: `Failed to load Static Resource Metadata: ${resourceMetadataPath}`
        });
      } else {
        throw error;
      }
    }
    return siteResources;
  }
  readStaticAssetsMetadata(staticRoot, metadataPath) {
    let assetMetadataPath;
    let siteAssets = {
      assets: {}
    };
    try {
      assetMetadataPath = import_path.default.join(staticRoot, metadataPath);
      siteAssets = import_fs_extra.default.readJSONSync(assetMetadataPath);
    } catch (error) {
      if (error.code === "ENOENT") {
        import_diagnostics.logger.debug({
          label: `SiteMetadata`,
          message: `Failed to load Static Resource Metadata: ${assetMetadataPath}`
        });
      } else {
        throw error;
      }
    }
    return siteAssets;
  }
};
function resolveStaticBundleVersion(metadataVersion, requestedVersion) {
  return metadataVersion || requestedVersion || import_shared_utils.VERSION_NOT_PROVIDED;
}
function parseSiteId(input) {
  const parts = input.split("|");
  const specifier = parts[0];
  const variants = {};
  for (let i = 1; i < parts.length; i++) {
    const [sigil, value] = parts[i].split("/");
    if (sigil && value) {
      variants[sigil] = value;
    } else if (sigil) {
      variants[sigil] = "true";
    }
  }
  return {
    specifier,
    variants
  };
}
function getSiteBundleId({specifier, namespace, name = "", version}, locale, ssr, i18n) {
  if (!specifier) {
    specifier = namespace ? `${namespace}/${name}` : name;
  }
  const versionedSpecifier = version && version !== import_shared_utils.VERSION_NOT_PROVIDED ? `${specifier}${SITE_VERSION_PREFIX}${(0, import_shared_utils.normalizeVersionToUri)(version)}` : specifier;
  const ssrSpecifier = ssr ? `${versionedSpecifier}${SITE_SSR_PREFIX}` : versionedSpecifier;
  return i18n?.defaultLocale === locale ? ssrSpecifier : `${ssrSpecifier}${SITE_LOCALE_PREFIX}${locale}`;
}
function getSiteResourceId({specifier, version}) {
  const versionedSpecifier = version && version !== import_shared_utils.VERSION_NOT_PROVIDED ? `${specifier}${SITE_VERSION_PREFIX}${(0, import_shared_utils.normalizeVersionToUri)(version)}` : specifier;
  return versionedSpecifier;
}
