// @lwrjs/lwc-ssr/build/cjs/__mocks__/undici.cjs
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};

// packages/@lwrjs/lwc-ssr/src/__mocks__/undici.ts
__markAsModule(exports);
__export(exports, {
  Pool: () => Pool
});
var Pool = class {
  constructor(origin) {
    this.origin = "";
    this.origin = origin;
  }
  async request(options) {
    const {path} = options;
    if (path.endsWith("bad"))
      throw new Error("fail");
    const ret = {origin: this.origin, req: options};
    const retStr = JSON.stringify(ret);
    return Promise.resolve({
      statusCode: path.endsWith("fail") ? 500 : path.endsWith("empty") ? 204 : path.endsWith("reset") ? 205 : 200,
      headers: {"cache-control": "no-cache", "X-something": void 0, "set-cookie": ["a", "b", "c"]},
      body: {
        arrayBuffer: () => Promise.resolve(Uint8Array.from(retStr.split("").map((letter) => letter.charCodeAt(0)))),
        text: () => Promise.resolve(retStr),
        json: () => Promise.resolve(ret)
      }
    });
  }
  close() {
  }
};
