// @lwrjs/server/build/cjs/express/express-response.cjs
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};

// packages/@lwrjs/server/src/express/express-response.ts
__markAsModule(exports);
__export(exports, {
  default: () => express_response_default
});
var ExpressResponse = class {
  constructor(expressResponse, isSiteGenerationRequest) {
    this.res = expressResponse;
    this.isSiteGenerationRequest = isSiteGenerationRequest || false;
  }
  hasHeader(name) {
    return this.res.hasHeader(name);
  }
  set(header) {
    this.res.set(header);
  }
  setHeader(name, value) {
    this.res.setHeader(name, value);
  }
  setSiteGenerationMetadata(meta) {
    if (this.isSiteGenerationRequest) {
      this.metadata = meta;
    }
  }
  status(status) {
    this.res.status(status);
    return this;
  }
  send(object) {
    if (this.isSiteGenerationRequest) {
      const context = {
        fs: {
          metadata: this.metadata
        }
      };
      this.res.send(context);
    } else {
      this.res.send(object);
    }
  }
  sendStatus(httpCode) {
    this.res.sendStatus(httpCode);
  }
  stream(stream) {
    if (this.isSiteGenerationRequest) {
      const context = {
        fs: {
          metadata: this.metadata
        }
      };
      stream.destroy();
      this.res.send(context);
    } else {
      stream.pipe(this.res);
    }
  }
  type(type) {
    this.res.type(type);
    return this;
  }
  cookie(name, value, options) {
    this.res.cookie(name, value, options);
    return this;
  }
};
var express_response_default = ExpressResponse;
