// @lwrjs/tools/build/cjs/server-build.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/tools/src/server-build.ts
__markAsModule(exports);
__export(exports, {
  REQUEST_SCRIPT_NAME: () => REQUEST_SCRIPT_NAME,
  buildServer: () => buildServer
});
var import_path = __toModule(require("path"));
var import_fs_extra = __toModule(require("fs-extra"));
var import_esbuild = __toModule(require("esbuild"));
var import_perf_hooks = __toModule(require("perf_hooks"));
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_config = __toModule(require("@lwrjs/config"));
var import_request_script = __toModule(require("./util/request-script.cjs"));
var import_generate_entry_plugin = __toModule(require("./plugins/generate-entry-plugin.cjs"));
var import_build_server_plugin = __toModule(require("./plugins/build-server-plugin.cjs"));
var REQUEST_SCRIPT_NAME = "index.js";
async function build(buildOptions, config) {
  const {outputDir, normalizedOutputDir, minify} = buildOptions;
  const ssrMetaContent = {
    minLwcVersion: import_config.MIN_LWC_VERSION,
    minLwrVersion: import_config.MIN_LWR_VERSION
  };
  const ssrMetaJsonContent = JSON.stringify(ssrMetaContent);
  await import_esbuild.default.build({
    entryPoints: ["./lwr.entry.js"],
    bundle: true,
    minify,
    sourcemap: true,
    sourcesContent: false,
    format: "cjs",
    platform: "node",
    logLevel: "silent",
    legalComments: "none",
    mainFields: ["module", "main"],
    external: [
      "node:*",
      "fsevents",
      "esbuild*",
      "esinstall",
      "@lwrjs/loader"
    ],
    banner: {
      js: createEnvVarHeader()
    },
    plugins: [(0, import_generate_entry_plugin.default)(), (0, import_build_server_plugin.default)(config, outputDir)],
    outfile: import_path.default.join(normalizedOutputDir, "ssr.js")
  });
  await import_fs_extra.default.writeFile(import_path.default.join(normalizedOutputDir, REQUEST_SCRIPT_NAME), (0, import_request_script.getMainScriptContent)());
  await import_fs_extra.default.writeFile(import_path.default.join(normalizedOutputDir, "ssr-meta.json"), ssrMetaJsonContent);
}
async function buildServer(configArg, options) {
  const startTime = import_perf_hooks.performance.now();
  const outputDir = options?.outputDir ?? "/build";
  const normalizedOutputDir = import_path.default.join(configArg?.rootDir || process.cwd(), outputDir);
  await import_fs_extra.default.ensureDir(normalizedOutputDir);
  await build({outputDir, normalizedOutputDir, minify: !!options?.minify}, configArg);
  const endTime = import_perf_hooks.performance.now();
  const timeDiff = endTime - startTime;
  import_diagnostics.logger.info({
    label: `buildServer`,
    message: `successfully built the server in ${Math.round(timeDiff)} ms`
  });
}
function createEnvVarHeader() {
  const currentFeatureFlags = (0, import_shared_utils.getFeatureFlags)();
  let initFeatureFlagsString = "";
  for (const [key, val] of Object.entries(currentFeatureFlags)) {
    if (val && val !== "false") {
      initFeatureFlagsString += `process.env.${key} ||= '${val.toString()}';`;
    }
  }
  const LWR_VERSION_ASSIGNMENT = `globalThis.LWR_VERSION='${import_config.LWR_VERSION}';`;
  const LWC_VERSION_ASSIGNMENT = `globalThis.LWC_VERSION='${import_config.LWC_VERSION}';`;
  const PWA_KIT_RUNTIME_VERSION_ASSIGNMENT = `globalThis.PWA_KIT_RUNTIME_VERSION='${import_config.PWA_KIT_RUNTIME_VERSION}';`;
  return `${LWR_VERSION_ASSIGNMENT}${LWC_VERSION_ASSIGNMENT}${PWA_KIT_RUNTIME_VERSION_ASSIGNMENT}${initFeatureFlagsString}`;
}
