// @lwrjs/server/build/es/express/express-request.js
import { AbstractMiddlewareRequest } from '../common/abstract-request.js';
export default class ExpressRequest extends AbstractMiddlewareRequest {
    constructor(expressRequest, isSiteGenerationRequest) {
        super(isSiteGenerationRequest);
        this.req = expressRequest;
    }
    get params() {
        return this.req.params;
    }
    get originalUrl() {
        return this.req.originalUrl;
    }
    get path() {
        return this.req.path;
    }
    get query() {
        return this.req.query;
    }
    get body() {
        // requires body parsing middleware
        return this.req.body;
    }
    get method() {
        return this.req.method;
    }
    get headers() {
        return this.req.headers;
    }
    get protocol() {
        return this.req.protocol;
    }
    get basePath() {
        return this.req.basePath || '';
    }
    set basePath(bp) {
        this.req.basePath = bp;
    }
    get locale() {
        return this.req.locale;
    }
    set locale(l) {
        this.req.locale = l;
    }
    acceptsLanguages() {
        return this.req.acceptsLanguages();
    }
    get(field) {
        return this.req.get(field);
    }
}
