// @lwrjs/server/build/es/express/express-response.js
export default class ExpressResponse {
    constructor(expressResponse, isSiteGenerationRequest) {
        this.res = expressResponse;
        this.isSiteGenerationRequest = isSiteGenerationRequest || false;
    }
    hasHeader(name) {
        return this.res.hasHeader(name);
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    set(header) {
        this.res.set(header);
    }
    setHeader(name, value) {
        this.res.setHeader(name, value);
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    setSiteGenerationMetadata(meta) {
        if (this.isSiteGenerationRequest) {
            this.metadata = meta;
        }
    }
    status(status) {
        this.res.status(status);
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    send(object) {
        if (this.isSiteGenerationRequest) {
            const context = {
                fs: {
                    metadata: this.metadata,
                },
            };
            this.res.send(context);
        }
        else {
            this.res.send(object);
        }
    }
    sendStatus(httpCode) {
        this.res.sendStatus(httpCode);
    }
    stream(stream) {
        if (this.isSiteGenerationRequest) {
            const context = {
                fs: {
                    metadata: this.metadata,
                },
            };
            // destroying the stream to prevent lingering processes
            stream.destroy();
            this.res.send(context);
        }
        else {
            stream.pipe(this.res);
        }
    }
    type(type) {
        this.res.type(type);
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    cookie(name, value, options) {
        this.res.cookie(name, value, options);
        return this;
    }
}
