// @lwrjs/server/build/es/koa/koa-request.js
import { AbstractMiddlewareRequest } from '../common/abstract-request.js';
export default class KoaRequest extends AbstractMiddlewareRequest {
    constructor(ctx, isSiteGenerationRequest) {
        super(isSiteGenerationRequest);
        this.ctx = ctx;
    }
    get params() {
        return this.ctx.params || {};
    }
    get originalUrl() {
        return this.ctx.request.originalUrl;
    }
    get path() {
        return this.ctx.request.path;
    }
    get query() {
        return this.ctx.request.query;
    }
    get protocol() {
        return this.ctx.protocol;
    }
    get body() {
        return this.ctx.body;
    }
    get method() {
        return this.ctx.method;
    }
    get headers() {
        return this.ctx.headers;
    }
    get basePath() {
        return this.ctx.basePath || '';
    }
    set basePath(bp) {
        this.ctx.basePath = bp;
    }
    get locale() {
        return this.ctx.locale;
    }
    set locale(l) {
        this.ctx.locale = l;
    }
    acceptsLanguages() {
        return this.ctx.request.acceptsLanguages();
    }
    get(field) {
        return this.ctx.get(field);
    }
    cookie(key) {
        return this.ctx.cookies.get(key);
    }
}
