// @lwrjs/server/build/es/koa/koa-response.js
import { PassThrough } from 'stream';
export default class KoaResponse {
    constructor(ctx, isSiteGenerationRequest) {
        this.ctx = ctx;
        this.isSiteGenerationRequest = isSiteGenerationRequest || false;
    }
    hasHeader(name) {
        return this.ctx.response.headers[name] !== undefined;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    set(header) {
        this.ctx.response.set(header);
    }
    setHeader(name, value) {
        this.ctx.response.set(name, value);
    }
    status(status) {
        this.ctx.response.status = status;
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    setSiteGenerationMetadata(meta) {
        if (this.isSiteGenerationRequest) {
            this.metadata = meta;
        }
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    send(object) {
        if (this.isSiteGenerationRequest) {
            const context = {
                fs: {
                    metadata: this.metadata,
                },
            };
            this.ctx.response.body = context;
        }
        else {
            this.ctx.response.body = object;
        }
    }
    sendStatus(httpCode) {
        this.ctx.response.status = httpCode;
    }
    stream(stream) {
        if (this.isSiteGenerationRequest) {
            const context = {
                fs: {
                    metadata: this.metadata,
                },
            };
            // destroying the stream to prevent lingering processes
            stream.destroy();
            this.ctx.response.body = context;
        }
        else {
            this.ctx.response.body = stream.on('error', this.ctx.onerror).pipe(new PassThrough());
        }
    }
    type(type) {
        this.ctx.type = type;
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    cookie(name, value, options) {
        this.ctx.cookies.set(name, value, options);
        return this;
    }
}
