// @lwrjs/tools/build/es/server-build.js
import path from 'path';
import fs from 'fs-extra';
import esbuild from 'esbuild';
import { performance } from 'perf_hooks';
import { logger } from '@lwrjs/diagnostics';
import { getFeatureFlags } from '@lwrjs/shared-utils';
import { LWR_VERSION, LWC_VERSION, PWA_KIT_RUNTIME_VERSION, MIN_LWR_VERSION, MIN_LWC_VERSION, } from '@lwrjs/config';
import { getMainScriptContent } from './util/request-script.js';
import generateLwrEntry from './plugins/generate-entry-plugin.js';
import buildLwrServer from './plugins/build-server-plugin.js';
export const REQUEST_SCRIPT_NAME = 'index.js';
async function build(buildOptions, config) {
    const { outputDir, normalizedOutputDir, minify } = buildOptions;
    const ssrMetaContent = {
        minLwcVersion: MIN_LWC_VERSION,
        minLwrVersion: MIN_LWR_VERSION,
    };
    const ssrMetaJsonContent = JSON.stringify(ssrMetaContent);
    // building specifically for MRT/Lambda
    await esbuild.build({
        entryPoints: ['./lwr.entry.js'],
        bundle: true,
        minify,
        sourcemap: true,
        sourcesContent: false,
        format: 'cjs',
        platform: 'node',
        logLevel: 'silent',
        legalComments: 'none',
        // Alow us to roll up additional require() dependencies
        mainFields: ['module', 'main'],
        external: [
            // node native dependencies
            'node:*',
            // fsevents used by chokidar used by nunjucks
            'fsevents',
            // These dependencies are not needed to serve a basic generated site, but they are still excluded
            // in case there is a need to add a custom module provider.
            'esbuild*',
            'esinstall',
            // We can remove this once we have the module bundler working
            '@lwrjs/loader',
        ],
        banner: {
            js: createEnvVarHeader(),
        },
        plugins: [generateLwrEntry(), buildLwrServer(config, outputDir)],
        // MRT expects an entry point named `ssr.js`
        outfile: path.join(normalizedOutputDir, 'ssr.js'),
    });
    // Save the request script to the app folder
    await fs.writeFile(path.join(normalizedOutputDir, REQUEST_SCRIPT_NAME), getMainScriptContent());
    await fs.writeFile(path.join(normalizedOutputDir, 'ssr-meta.json'), ssrMetaJsonContent);
}
/**
 * Resolve configurations, generate a server build module, and bundle
 * the LWR server.  The bundled server will be written to the file system.
 *
 * @param {LwrGlobalConfig} configArg - programmatic LWR global config
 * @param {BuildOptions} options - server build options
 */
export async function buildServer(configArg, options) {
    const startTime = performance.now();
    const outputDir = options?.outputDir ?? '/build';
    const normalizedOutputDir = path.join(configArg?.rootDir || process.cwd(), outputDir);
    await fs.ensureDir(normalizedOutputDir);
    await build({ outputDir, normalizedOutputDir, minify: !!options?.minify }, configArg);
    const endTime = performance.now();
    const timeDiff = endTime - startTime;
    logger.info({
        label: `buildServer`,
        message: `successfully built the server in ${Math.round(timeDiff)} ms`,
    });
}
/**
 * Build in important environment variables set during build
 */
function createEnvVarHeader() {
    // Setup global environment variables
    const currentFeatureFlags = getFeatureFlags();
    let initFeatureFlagsString = '';
    for (const [key, val] of Object.entries(currentFeatureFlags)) {
        if (val && val !== 'false') {
            // Allow server env vars take precedence via ||=
            initFeatureFlagsString += `process.env.${key} ||= '${val.toString()}';`;
        }
    }
    const LWR_VERSION_ASSIGNMENT = `globalThis.LWR_VERSION='${LWR_VERSION}';`;
    const LWC_VERSION_ASSIGNMENT = `globalThis.LWC_VERSION='${LWC_VERSION}';`;
    const PWA_KIT_RUNTIME_VERSION_ASSIGNMENT = `globalThis.PWA_KIT_RUNTIME_VERSION='${PWA_KIT_RUNTIME_VERSION}';`;
    return `${LWR_VERSION_ASSIGNMENT}${LWC_VERSION_ASSIGNMENT}${PWA_KIT_RUNTIME_VERSION_ASSIGNMENT}${initFeatureFlagsString}`;
}
